src/blocked.c

```diff
@@ -111,6 +103,8 @@ void blockClient(client *c, int btype) {
 void updateStatsOnUnblock(client *c, long blocked_us, long reply_us, int had_errors){
```

Comment:
let's comment that this method is only used on timeout, error and modules, and not on most unblocked cases.
and in that light, i wonder if we wanna add them to the monitor (i.e. the timeout cases)?

Reply:
Since this is currently not only a case of timeout (modules run this command as well) I do not think we can avoid reporting it to monitors.

Reply:
i meant that we can add an argument or something so we can make a distinction between timeout / error and modules.

Reply:
So you are O.K with totaly not reporting timedout commands?

Reply:
if we log MONITOR when a command was actually executed, and the command never did, i think it makes sense not to log it.
it's the same as a command that got rejected by OOM or other error.
maybe we need another pair of eyes on this.
@itamarhaber @madolson ?

Reply:
Just to place my personal opinion:
I feel that the monitor should log every command once it is being handled. even in case a command was timedout, it does not mean it was not handled. I can find some reason to monitor both commands which are blocked and also monitor again the fact that they are reprocessed (maybe add some indication in the monitor output) 
But I feel it can be confusing not monitor blocked commands which were timedout.

Reply:
some day, MONITOR will also be able to track commands that failed (e.g. on OOM) and so on.
we have great plans for MONITOR in the far future, see https://github.com/redis/redis/issues/6321#issuecomment-1175062204
in the meanwhile, with the current monitor, i don't think it should log things that didn't take place.
i.e. if you'll see a BLPOP in the monitor, you'll think an element was popped, when in fact the command failed.
i.e. it failed in a similar manner as an OOM, etc.

it's true that it it would have failed on WRONGTYPE we would have logged it anyway, but in my eyes, that's just because the current infra is limited.

Reply:
I strongly agree with Ran - in principle, MONITOR should log every command as it is handled.

Errors and timeouts aren't the same things IMO, although both may be triggered by circumstances/topology and the dataset.

Because timeouts are asynchronous by definition, the only consistent way to treat the potential ones (i.e. all blocking commands) that I see is to log them as they are first processed (which appears to be the current behavior).

Because MONITOR is less-than-ideal for replay purposes, if that's needed, I feel a new TRACE command, or perhaps some variation on AOF, is better.

Reply:
It may be the wine speaking, but I don't understand any of this.
Isn't monotor used to track which commands were actually executed, like a live AOF stream that also includes read commands? it does not include commands that were rejected (OOM). how is it useful for the user to log timed out commands ?

Reply:
You raise a good point. I guess that it should be clear if the monitor command should report commands which were "intercepted" or actually "processed".
Currently our documentation indeed states the later, however the current implementation indicates the first. My point was that "timedout" commands are a different case than "rejected" commands. Currently we do not report commands which failed authorization (and OOM), but a timedout command is actually being processed and the fact that it was timedout does not mean it was not handled (it does mean it did not fully cause the requested effect). For example let's say a command to read from a stream with non existing group was received, we will report it on the monitor even though it ended with error. I think of timeout as an error to complete processing the command, unlike the case of a command which was rejected.

Reply:
the example you gave about stream with non existing group is similar to the one i gave earlier about WRONGTYPE.
i think these cause less damage to the MONITOR user than reporting a command that failed on timeout.
I think some use cases for monitor will try to feed the command stream back to redis, in the case of WRONGTYPE, the command will consistently fail again immediately, but in the case of timeout, i could either hang, or worse, succeed for some reason.
i'll try to grab @yossigo to comment here as well.

Reply:
i discussed it with Yossi, and he convinced me to flip this around.
the arguments were:
1. we were trying to conclude if MONITOR is about the database (which command modified the data), or about clients (which client executed a command). for the first, there are already ways to "monitor" the database, which are keyspace notification, AOF, and attaching a replica, but for the later, there's nothing so many monitor should serve this purpose (monitor client commands and not database changes).
2. we have some future plans for monitor (https://github.com/redis/redis/issues/6321#issuecomment-1175062204) so some day it'll be able to also log rejected command, and maybe also both the being blocked time and the being released time, as well as the release reason (timeout or not), so maybe for now, the right thing to do is to keep it was it was and avoid any logical change (other than clear bug fixes), so we don't end up incrementally changing it and breaking some use case multiple times.

so the bottom line of that discussion is that in contrast to my previous opinion (to make a change to avoid logging on timeouts), i'd like to revert an earlier change we made, and keep logging blocked clients in the initial `call()`, when being blocked.

Reply:
@oranagra I implemented according to the decision to keep monitors current behavior. 
I placed a flag which indicates a client is currently executing a command which is cleared once the cmd->proc is completed and the client is no longer blocked.
The main issue with this is that it will not work well in case of nested calls (LUA, multi etc...) but since currently we do not allow blocking in nested calls (I am still not 100% sure regarding modules) this should be fine.